File: 백준/Gold/22945. 팀 빌딩/팀 빌딩.py
"""
투 포인터를 이용하자.
양 끝점의 최솟값이 최대가 되도록 투 포인터를 움직이자.

[1차 채점] WA.
반례) 150 50 30 50 150 1 10
투 포인터 중 값이 더 작은 것을 버리도록 이동하는게 맞겠다.
"""
import sys


def input():
    return sys.stdin.readline().rstrip()


def main():
    n = int(input())
    stats = list(map(int, input().split()))

    lp, rp = 0, n - 1
    max_res = 0
    while lp < rp:
        max_res = max(max_res, (rp - lp - 1) * min(stats[lp], stats[rp]))
        if stats[lp] > stats[rp]:
            rp -= 1
        else:
            lp += 1
    print(max_res)


if __name__ == '__main__':
    main()

File: 백준/Gold/22945. 팀 빌딩/팀 빌딩.java
// s = 0, e = N - 1 인 s, e 에 대해
// stats[s] > stats[e] 가 성립한다고 하자
// e의 경우 s와 매칭이 되었을 때 유일한 최적이므로
// 이 값을 정답에 반영하고 나면 e의 상대를 더 이상 탐색할 필요가 없다
// 따라서 두 포인터 중 능력치가 낮은 포인터를 이동시키면서 정답을 탐색하면 된다

// 그렇다면 s > 0, e < N - 1 인 s, e 에 대해서는
// stats[s] > stats[e] 가 성립한다고 할 때
// e의 경우 s와 매칭이 되었을 때 유일한 최적이라고 말할 수 있을까?
// 그렇게 말할 수 있는지 여부는 몰라도,
// 구간 [0, s - 1]에 속하는 모든 포인터는 자신이 작은 값이 되어 최적의 값으로서 계산에 반영된 적이 있기 때문에
// e의 상대로 [0, s - 1]에 속하는 포인터들을 고려할 필요는 더 이상 없다


import java.io.*;
import java.util.*;


public class Main {

    static int N;
    static int[] stats;

    public static void main(String[] args) throws Exception {
        // 입력을 받는다
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        stats = new int[N];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            stats[i] = Integer.parseInt(st.nextToken());
        }

        // 투 포인터를 이용해 정답을 갱신한다
        int answer = 0;
        int a = 0;
        int b = N - 1;
        while (a < b) {
            int statA = stats[a];
            int statB = stats[b];
            answer = Math.max(answer, (b - a - 1) * Math.min(statA, statB));
            if (statA == statB) {
                a++;
                b--;
            } else if (statA > statB) {
                b--;
            } else {
                a++;
            }
        }

        // 정답을 출력한다
        System.out.println(answer);
    }
}
